### PR TITLE
Feat/criteria reason

### PR DESCRIPTION
**Pull request by jlarfors**
Another significant PR. This is for support for the release criteria.
It meant updating the `criteria` resource, and to make the GraphQL queries work with more ease, some minor changes to the `store`.

### CODE REVIEW

File: api/core/resource.go [hunk 1]
```diff
@@ -25,6 +25,12 @@ type ResourceBlockHCLWrapper struct {
 	ResourceBlock ResourceBlock `hcl:"resource,block"`
 }
 
+// ResourceBlockHCLWrapper is a simple wrapper for a ResourceBlock which is
+// used when decoding a single resource block in HCL/JSON.
+type ResourceBlockJSONWrapper struct {
```
Comment (olliefr) [inline api/core/resource.go:30]: Comment and type name don't match.

File: api/v1/criteria.go [hunk 2]
```diff
@@ -1,147 +1,177 @@
...
-		resID := fmt.Sprintf("%s/%s", string(core.QueryResourceKind), querySpec.Name)
-		resource, output := common.RunResource(bCtx, ctx, resID, cty.NilVal)
-		if output.Error != nil {
+
+	queryVal, err := c.queryToCtyValue(bCtx, ctx)
+	if err != nil {
+		return core.ResourceOutput{
+			ID:     c.ID(),
+			Status: events.ResourceRunFailure,
+			// The returned error should contain enough context about what went wrong
```
Comment (olliefr) [inline api/v1/criteria.go:52]: Not sure this comment adds value?

File: bubbly/builtin/release.go [hunk 3]
```diff
@@ -1,18 +1,32 @@
...
-			if len(criteria.ReleaseEntry) == 0 {
-				return "PENDING"
-			}
-			for _, entry := range criteria.ReleaseEntry {
-				if !entry.Result {
-					return "BLOCKED"
-				}
-			}
+		status := ReleaseStageStatus(stage)
+		if status != "READY" {
```
Comment (olliefr) [inline bubbly/builtin/release.go:6]: Maybe `iota` with a `String()` method for these string constants? [Example](https://yourbasic.org/golang/iota/)

File: bubbly/release_get.go [hunk 4]
```diff
@@ -86,7 +92,7 @@ func GetRelease(bCtx *env.BubblyContext) (*builtin.Release, error) {
 	}
 
 	if len(releases.Release) == 0 {
-		return nil, fmt.Errorf("release does not exist")
+		return nil, ErrReleaseNotExist
```
Comment (olliefr) [inline bubbly/release_get.go:95]: I don't think we've been using this error handling/reporting pattern anywhere else? Just checking.

File: cmd/util/release.go [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+		yellow = color.New(color.FgYellow).SprintFunc()
+		red    = color.New(color.FgRed).SprintFunc()
+	)
+
+	switch status {
+	case "BLOCKED":
+		return red(status)
+	case "PENDING":
+		return yellow(status)
+	case "READY":
```
Comment (olliefr) [inline cmd/util/release.go:18]: As I said in another comment, maybe consider introducing enumeration type for these constants?

File: store/postgres_graphql.go [hunk 6]
```diff
@@ -116,20 +116,26 @@ func psqlResolveRootQuery(pool *pgxpool.Pool, tenant string, graph *SchemaGraph,
...
+	var hasRows bool
 	for rows.Next() {
+		hasRows = true
 		if err := psqlScanRowColumns(rows, result, rootColumns); err != nil {
 			return nil, fmt.Errorf("failed scanning row values: %w", err)
 		}
 	}
+	if !hasRows {
+		// Initialize with an empty slice to avoid returning just null
+		result[rootTable] = make([]interface{}, 0)
```
Comment (olliefr) [inline store/postgres_graphql.go:133]: Can this not be done _before_ entering the loop? I presume if the code inside the loop runs, the `result[rootTable]` would be re-initialised or replaced anyway?

File: store/postgres_graphql.go [hunk 7]
```diff
@@ -289,6 +301,11 @@ func psqlSubQuery(tenant string, graph *SchemaGraph, qb *sqlQueryBuilder, root *
 		}
 	}
 
+	// By default if no order_by specified, order by the primary key in DESC order
+	if orderByArg == nil && qb.depth == 1 {
+		qb.sql = qb.sql.OrderBy(tableColumn(tc.alias, tableIDField) + " DESC")
```
Comment (olliefr) [inline store/postgres_graphql.go:306]: There are string constants defined for the sorting order: `orderAsc` and `orderDesc` at the top of this file :)

**Review by olliefr — APPROVED** [review]
LGTM

File: bubbly/builtin/release.go (hunk 3 above)
Comment (jlarfors) [reply bubbly/builtin/release.go:6]: Yes I want to do that once I know what the values are and where to put them

Comment (jlarfors) [reply bubbly/builtin/release.go:6]: Done

File: store/postgres_graphql.go (hunk 6 above)
Comment (jlarfors) [reply store/postgres_graphql.go:133]: Nice suggestion! Fixed

**Pull request merged**